[OSDOCS#18861: Added a note for sos report (#110821)
openshift/openshift-docs: modules/installation-manual-recovering-when-auto-recovery-is-unavail.adoc

$ oc edit secret <node_name>-fencing
----
+
If the cluster recovers after updating the fencing secrets, you do not need to take further action. If the issue persists, proceed to the next step.

. Recover from a single-node failure:

> .. Gather initial diagnostics.

Reviewer: Should there be an example here so a user knows what they are looking at ? That displays a lot of information, but not all of it is helpful in diagnosing an issue.

Reviewer 2: Can you clarify what you mean here? do you want more examples of all the diagnostic commands below?

Reviewer: Should the section `Full List of Resources` be placed here as an example of the state of the cluster or would that information not be useful for a user ?

Reviewer 2: I feel like there's a lot of information. Maybe we want to specifically highlight the relevant parts sudo pcs status when it's relevant or pacemakercluster when it's relevant. Giving them a full overview feels like a lot of information. What about this as a shorthand?

```
[ec2-user@aws-jpoulin-dev ~]$ oc get pacemakercluster/cluster -o json | jq -r '
    "Cluster: Healthy=\(.status.conditions[] | select(.type=="Healthy") | .status)",
    (.status.nodes[] |
      "Node \(.nodeName): Healthy=\(.conditions[] | select(.type=="Healthy") | .status)",
      (.fencingAgents[] | "  Fencing \(.name): Healthy=\(.conditions[] | select(.type=="Healthy") | .status)"),
      (.resources[] | "  Resource \(.name): Healthy=\(.conditions[] | select(.type=="Healthy") | .status)")
    )
  '
Cluster: Healthy=True
Node master-0: Healthy=True
  Fencing master-0_redfish: Healthy=True
  Resource Kubelet: Healthy=True
  Resource Etcd: Healthy=True
Node master-1: Healthy=True
  Fencing master-1_redfish: Healthy=True
  Resource Kubelet: Healthy=True
  Resource Etcd: Healthy=True
```

Reviewer 2: Here are handy ways to get "wide" statuses for each level:
```
● # Cluster-level conditions
  oc get pacemakercluster/cluster -o jsonpath='Cluster: {range .status.conditions[*]}{.type}={.status} {end}{"\n"}'

  # Node conditions
  oc get pacemakercluster/cluster -o jsonpath='{range .status.nodes[*]}{.nodeName}{": "}{range .conditions[*]}{.type}={.status}{" "}{end}{"\n"}{end}'

  # Kubelet resource conditions
  oc get pacemakercluster/cluster -o jsonpath='{range .status.nodes[*]}{.nodeName}{" Kubelet: "}{range .resources[?(@.name=="Kubelet")].conditions[*]}{.type}={.status}{" "}{end}{"\n"}{end}'

  # Etcd resource conditions
  oc get pacemakercluster/cluster -o jsonpath='{range .status.nodes[*]}{.nodeName}{" Etcd: "}{range .resources[?(@.name=="Etcd")].conditions[*]}{.type}={.status}{" "}{end}{"\n"}{end}'

  # Fencing agent conditions
  oc get pacemakercluster/cluster -o jsonpath='{range .status.nodes[*]}{.nodeName}{" "}{range .fencingAgents[*]}{.name}{": "}{range .conditions[*]}{.type}={.status}{" "}{end}{end}{"\n"}{end}'
```

Reviewer 2: The corresponding pacemaker commands would be:
```
# Cluster & Node List
sudo pcs cluster status

# Fencing Agents
sudo pcs stonith status

# Kubelet
sudo pcs resource status kubelet

# Etcd
sudo pcs resource status etcd
```

Reviewer 2: @bjahagir-OpenShift this whole section kind of sucks. I'd recommend we focus this PR on minimal changes that just address the core change we're trying to make - adding a link to the sos-report docs. all this context about information gathering we can refactor with the guide I provided you that explains how to gather information during the general debugging section.

.. Run the following additional diagnostic commands, if necessary:
+
Reset the resources on your cluster and instruct Pacemaker to attempt to start them fresh.
+
[source,terminal]
----
$ sudo pcs resource cleanup
----
+
Review all Pacemaker activities on the node.
+
[source,terminal]
----
$ sudo journalctl -u pacemaker
----
+
Diagnose etcd resource startup issues.
+
[source,terminal]
----
$ sudo journalctl -u pacemaker | grep podman-etcd
----

.. View the fencing configuration for the node.
+
[source,terminal]
----
$ sudo pcs stonith config <node_name>_redfish
----
+

> If fencing is not working, verify that the Redfish endpoint is accessible and that the credentials are correct.

Reviewer: What does the failure look like ? Another example here might help.

Reviewer 2: There are many ways this could present. I just disrupted vbmc & sushytools. looking at the API, I can see this via:
```
[ec2-user@aws-jpoulin-dev ~]$ oc get pacemakercluster/cluster -o jsonpath='{range .status.nodes[*]}{.nodeName}{" "}{range .fencingAgents[*]}{.name}{": "}{range .conditions[*]}{.type}={.status}{" "}{end}{end}{"\n"}{end}'
master-0 master-0_redfish: InService=True Managed=True Enabled=True Operational=False Active=True Started=False Schedulable=True Healthy=False 
master-1 master-1_redfish: InService=True Managed=True Enabled=True Operational=False Active=True Started=True Schedulable=True Healthy=False 
```
Notice both my fencing agents are no longer healthy. This will eventually bubble up to the cluster-etcd-operator status, but you can also see a summary like so:
```
[ec2-user@aws-jpoulin-dev ~]$   oc get etcd cluster -o json | jq -r '.status.conditions[] | select(.type=="PacemakerHealthCheckDegraded") | "PacemakerHealthCheckDegraded=\(.status) Reason=\(.reason) Message=\(.message)"'                                       
PacemakerHealthCheckDegraded=True Reason=PacemakerUnhealthy Message=master-0 node is unhealthy: fencing unavailable (no agents running); master-1 node is unhealthy: fencing unavailable (no agents running)
```

Eventually, this will present like so:
```
[ec2-user@aws-jpoulin-dev ~]$ oc get co/etcd
NAME   VERSION       AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
etcd   4.22.0-rc.2   True        False         True       121m    PacemakerHealthCheckDegraded: master-0 node is unhealthy: fencing unavailable (no agents running); master-1 node is unhealthy: fencing unavailable (no agents running)
```

Here's the problem, an actual fencing **failure** means you don't have an API to query anymore. So you're going to have to get that information from pcs from the nodes.

Example - starting from the unhealthy state above, I killed one of my control-plane nodes.
Then, I hopped onto the survivor and ran:
```
[core@master-0 ~]$ sudo pcs status 
Cluster name: TNF
Cluster Summary:
  * Stack: corosync (Pacemaker is running)
  * Current DC: master-0 (version 2.1.10-2.el9-5693eaeee) - partition with quorum
  * Last updated: Tue May  5 19:53:23 2026 on master-0
  * Last change:  Tue May  5 19:52:45 2026 by root via root on master-0
  * 2 nodes configured
  * 6 resource instances configured

Node List:
  * Node master-1: UNCLEAN (offline)
  * Online: [ master-0 ]

Full List of Resources:
  * Clone Set: kubelet-clone [kubelet]:
    * kubelet	(systemd:kubelet):	 Started master-1 (UNCLEAN)
    * Started: [ master-0 ]
  * master-0_redfish	(stonith:fence_redfish):	 FAILED master-0
  * master-1_redfish	(stonith:fence_redfish):	 FAILED master-0
  * Clone Set: etcd-clone [etcd]:
    * etcd	(ocf:heartbeat:podman-etcd):	 Started master-1 (UNCLEAN)
    * etcd	(ocf:heartbeat:podman-etcd):	 FAILED master-0

Failed Resource Actions:
  * master-0_redfish start on master-0 returned 'error' at Tue May  5 19:53:18 2026 after 1.328s
  * kubelet 1m-interval monitor on master-0 returned 'not running' (inactive) at Tue May  5 18:43:20 2026
  * etcd 30s-interval monitor on master-0 returned 'error' (master-0 must force a new cluster) at Tue May  5 19:52:45 2026
  * master-1_redfish start on master-0 returned 'error' at Tue May  5 19:53:17 2026 after 2.021s

Failed Fencing Actions:
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:22.165230Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:17.372190Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:12.597151Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:07.362094Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:02.907044Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:59.570007Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:53.680942Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:50.768909Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:44.022834Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:37.945767Z
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:32.057702Z

Daemon Status:
  corosync: active/enabled
  pacemaker: active/enabled
  pcsd: active/enabled
```
The key value here is:
```
Failed Fencing Actions:
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 
```
If fencing actions fail, they'll be in this status. Now, but simply restarting sushy-tools and vbmc, I can just wait, and then:
```
[core@master-0 ~]$ sudo pcs status
Cluster name: TNF
Cluster Summary:
  * Stack: corosync (Pacemaker is running)
  * Current DC: master-0 (version 2.1.10-2.el9-5693eaeee) - partition with quorum
  * Last updated: Tue May  5 19:55:28 2026 on master-0
  * Last change:  Tue May  5 19:55:06 2026 by root via root on master-0
  * 2 nodes configured
  * 6 resource instances configured

Node List:
  * Node master-1: pending
  * Online: [ master-0 ]

Full List of Resources:
  * Clone Set: kubelet-clone [kubelet]:
    * Started: [ master-0 ]
    * Stopped: [ master-1 ]
  * master-0_redfish	(stonith:fence_redfish):	 Started master-0
  * master-1_redfish	(stonith:fence_redfish):	 Started master-0
  * Clone Set: etcd-clone [etcd]:
    * etcd	(ocf:heartbeat:podman-etcd):	 FAILED master-0
    * Stopped: [ master-1 ]

Failed Resource Actions:
  * master-0_redfish start on master-0 returned 'error' at Tue May  5 19:54:55 2026 after 1.066s
  * kubelet 1m-interval monitor on master-0 returned 'not running' (inactive) at Tue May  5 18:43:20 2026
  * etcd 30s-interval monitor on master-0 returned 'error' (master-0 must force a new cluster) at Tue May  5 19:52:45 2026
  * master-1_redfish start on master-0 returned 'error' at Tue May  5 19:54:56 2026 after 1.549s

Failed Fencing Actions:
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:59.623072Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:55.442034Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:52.795010Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:49.127977Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:42.950921Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:38.744882Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:34.913847Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:31.267814Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:27.119777Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:23.496744Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:19.829710Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:15.724674Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:11.608640Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:54:03.207570Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:58.997535Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:54.862501Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:51.081470Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:47.412439Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:43.153404Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:40.443382Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:36.300347Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:32.090312Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:29.482291Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:22.165230Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:17.372190Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:12.597151Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:07.362094Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:53:02.907044Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:59.570007Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:53.680942Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:50.768909Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:44.022834Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:37.945767Z (a later attempt from master-0 succeeded)
  * reboot of master-1 by master-0 for pacemaker-controld.54319@master-0 last failed at 2026-05-05 19:52:32.057702Z (a later attempt from master-0 succeeded)

Daemon Status:
  corosync: active/enabled
  pacemaker: active/enabled
  pcsd: active/enabled
```
The key difference you'll notice here is "a later attempt from master-0 succeeded". This means that those records are now historical, because recovery was able to proceed.

Reviewer 2: After a minute, my cluster is healthy again:
```
[core@master-0 ~]$ sudo pcs resource status
  * Clone Set: kubelet-clone [kubelet]:
    * Started: [ master-0 master-1 ]
  * Clone Set: etcd-clone [etcd]:
    * Started: [ master-0 master-1 ]
```
After a few minutes, the health check will update as well:
```
[ec2-user@aws-jpoulin-dev ~]$   oc get etcd cluster -o json | jq -r '.status.conditions[] | select(.type=="PacemakerHealthCheckDegraded") | "PacemakerHealthCheckDegraded=\(.status) Reason=\(.reason) Message=\(.message)"'                                       
PacemakerHealthCheckDegraded=False Reason=null Message=null
```

Reviewer 2: Part of docs refactor, not relevant to this PR.

> .. If fencing is working correctly but etcd does not start correctly, recover etcd by restoring from a backup:

Reviewer: Is this the correct way to restore from backup ? This seems incorrect @jaypoulz

Reviewer 2: This section should link to the restoring etcd from backup part of the docs, when that part exists. I agree that the command below is not sufficient.

Reviewer 2: Ideally should be part of the docs refactor; not relevant to this PR.

Author: Just need to add xref for backup etcd data

$ sudo chown -R etcd:etcd /var/lib/etcd
----
+
If the recovery is successful, you do not need to take further action. If the issue persists, proceed to the next step.

. Recover from a complete node failure:

.. Power on both control plane nodes.
+
When Pacemaker detects that both nodes are online, it starts automatically and begins the recovery operation. If the recovery does not start as expected, use the diagnostic commands described in the earlier step to investigate the issue.

.. Clean up the cluster resources:
+
[source,terminal]
----
$ sudo pcs resource cleanup
----

.. Check resource start order.
+
[source,terminal]
----
$ sudo pcs status --full
----

> .. Inspect the Pacemaker service journal if kubelet fails by running the following commands:

Reviewer: There does not seem like there is enough context. What errors would a user see ?

Reviewer 2: Easiest way to detect kubelet failures.

**Using oc to query PacemakerCluster**
_This is the status as reported from inside k8s. Healthy is the primary status, the others explain what is going on in pacemaker._
Command
```
$ oc get pacemakercluster/cluster -o jsonpath='{range .status.nodes[*]}{.nodeName}{" Kubelet: "}{range .resources[?(@.name=="Kubelet")].conditions[*]}{.type}={.status}{" "}{end}{"\n"}{end}'
```
Output:
```
master-0 Kubelet: InService=True Managed=True Enabled=True Operational=True Active=True Started=True Schedulable=True Healthy=True 
master-1 Kubelet: InService=True Managed=True Enabled=True Operational=True Active=True Started=True Schedulable=True Healthy=True
```

**Using PCS from one of the nodes**
Command:
```
sudo pcs resource status kubelet
```
Healthy
```
  * Clone Set: kubelet-clone [kubelet]:
    * Started: [ master-0 master-1 ]
```
Failure
```
  * Clone Set: kubelet-clone [kubelet]:
    * kubelet	(systemd:kubelet):	 FAILED master-0
    * Started: [ master-1 ]
```

Reviewer 2: Not relevant to current PR. Let's revisit this in refactor PR

$ sudo journalctl -u kubelet
----

.. Resolve etcd synchronization issues. 
+

> If one node has a more up-to-date etcd, Pacemaker attempts to fence the lagging node and start it as a learner. If the fencing process does not complete, verify the Redfish fencing endpoint and credentials by running the following command:

Reviewer: How will I know the fencing process does not complete ?

Reviewer 2: As I showed in the comment above, the only way to see this is using the `sudo pcs status` from one of the nodes and obversing the failed fencing action that are *not* annotated with "a later fencing action succeeded"

See above for what failed fencing looks like
https://github.com/openshift/openshift-docs/pull/110821#discussion_r3191192950

Reviewer 2: Let's revisit this with refactor PR.

+
[source,terminal]
----
$ sudo pcs stonith config
----
+
If the recovery is successful, you do not need to take further action. If the issue persists, perform manual recovery as described in the next step.

. If you need to manually recover from an event when one of the nodes is not recoverable, follow the procedure in "Replacing control plane nodes in a two-node OpenShift cluster with fencing".
+

> When a cluster loses a single node, it enters in `degraded` mode. In this state, Pacemaker automatically unblocks quorum and allows the cluster to temporarily operate on the remaining node.

Reviewer 2: drop the `in`. We just call this degraded-mode

> When a cluster loses a single node, it enters in `degraded` mode. In this state, Pacemaker automatically unblocks quorum and allows the cluster to temporarily operate on the remaining node.

Reviewer: Perhaps this should say something like 

```suggestion
When one of the two nodes fails, the cluster enters `degraded` mode. In this state, Pacemaker automatically unblocks quorum and allows the cluster to temporarily operate on the remaining node.
```

Author: Implement this

+
If both nodes fail, you must restart both nodes to restore quorum so that Pacemaker can resume normal cluster operations.
+
If you can restart only one of the two nodes, follow the node replacement procedure to manually restore quorum on the surviving node.
+
If manual recovery fails, collect a must-gather and sosreport, and file a bug.
+
[NOTE]
====

> When troubleshooting or reporting issues, collect an sosreport from both control plane nodes. The sosreport includes the Pacemaker log (`/var/log/pacemaker/pacemaker.log`), which is useful for diagnosing fencing and quorum issues.

Reviewer 2: can sosreport link to the section?

$ sudo pcs stonith config
----
+
If the recovery is successful, you do not need to take further action. If the issue persists, perform manual recovery as described in the next step.

. If you need to manually recover from an event when one of the nodes is not recoverable, follow the procedure in "Replacing control plane nodes in a two-node OpenShift cluster".
+
When a cluster loses a single node, it enters the degraded mode. In this state, Pacemaker automatically unblocks quorum and allows the cluster to temporarily operate on the remaining node.
+
If both nodes fail, you must restart both nodes to reestablish quorum so that Pacemaker can resume normal cluster operations.
+
If you can restart only one of the two nodes, follow the node replacement procedure to manually reestablish quorum on the surviving node.
+
If manual recovery is still required and it fails, collect a must-gather and SOS report, and file a bug.
+
[NOTE]
====

> When troubleshooting or reporting issues, collect an sosreport from both control plane nodes. The sosreport automatically includes the Pacemaker log. You need this log for diagnosing fencing and cluster quorum issues. 

Reviewer 2: I think this is fine, but it could be good to note that the log lives at `/var/log/pacemaker/pacemaker.log`
Maybe that's too much detail
